pkg/app/api/grpcapi/piped_api.go

```diff
@@ -963,6 +966,93 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
 	return nil, status.Errorf(codes.Unimplemented, "ReportUnregisteredApplicationConfigurations is not implemented yet")
 }
 
+// CreateDeploymentChain creates a new deployment chain object and all required commands to
+// trigger deployment for applications in the chain.
+func (a *PipedAPI) CreateDeploymentChain(ctx context.Context, req *pipedservice.CreateDeploymentChainRequest) (*pipedservice.CreateDeploymentChainResponse, error) {
+	projectID, _, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	findNodeApps := func(filter *pipedservice.CreateDeploymentChainRequest_ApplicationsFilter) ([]*model.Application, error) {
+		filters := []datastore.ListFilter{
+			{
+				Field:    "ProjectId",
+				Operator: datastore.OperatorEqual,
+				Value:    projectID,
+			},
+		}
+
+		if filter.AppName != "" {
+			filters = append(filters, datastore.ListFilter{
+				Field:    "Name",
+				Operator: datastore.OperatorEqual,
+				Value:    filter.AppName,
+			})
+		}
+
+		// TODO: Support find node apps by appKind and appLabels.
+
+		apps, _, err := a.applicationStore.ListApplications(ctx, datastore.ListOptions{
+			Filters: filters,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return apps, nil
+	}
+
+	chainNodes := make([]*model.DeploymentChainNode, 0, len(req.Filters))
+	apps := make([]*model.Application, 0)
+	for i, filter := range req.Filters {
+		nodeApps, err := findNodeApps(filter)
+		if err != nil {
+			return nil, err
+		}
+
+		apps = append(apps, nodeApps...)
+		chainNodes = append(chainNodes, &model.DeploymentChainNode{
+			Index:    int32(i + 1),
+			Runnable: false,
+		})
+	}
+
+	dc := model.DeploymentChain{
+		Id:        uuid.New().String(),
```

Comment:
Maybe we should add this chain ID to the deployment model.

Reply:
You're right, we need that as a reference to display which chain this deployment belongs to in the deployment detailed page 👍

Reply:
addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/050e72b0f1063021c77c4855c113db657e515cae

Reply:
The first deployment needs this ID as well. Do you think that we should create it by the first Piped?

Reply:
I see, in such case, the first piped will create deploymentChain object and register it to the control-plane just as we do in `trigger.triggerDeployment` currently, right?

Reply:
How it can create the whole deployment chain object while it doesn't have enough information of applications from other Piped?
Do we have any good ideas? 🤔

Reply:
No, we don't have to create a whole deployment chain object, just the deploymentChainId is enough, we still submit the filters used to find application and make command as we do in this PR, the deploymentChain object created by the piped just to enable us to use the deploymentChainId to trigger the first deployment of the chain, everything else in this PR has remained.

Reply:
👍

Reply:
How do you think about adding a chain option to `CreateDeploymentRequest` and initializing chain commands in that API instead of creating this new `CreateDeploymentChain`?

In that way,  we can have both chain ID and the first deployment ID at the same time.

Reply:
So you mean to make `CreateDeploymentRequest` has the ability to make chain commands to trigger all other applications in the chain and the RPC call to create deployment for the first application will do it? In that case, I wonder 2 things:
- we also need to provide matchers to filter out applications for chain block in that request body, right?
- though it's named `CreateDeployment`, it may also create `Command` in some conditions.

Besides, not sure what do you mean by "we can have both chain ID and the first deployment ID at the same time", since we separate the RPC to create the first deployment of the chain with the deployment chain itself, we have to first: create the deployment chain and wait until it's succeeded, then: use the deployment chain id created to find the deployment chain object in the datastore and add the on going create deployment as the deploymentChain.block[0] item, right?
Currently, that part is addressed here: https://github.com/pipe-cd/pipe/pull/2815/files#diff-dc8d28c3b055fcad0f022406b744ce756e3be8c630b6c40aef12e026c6a58dcfR1019-R1032

Reply:
> we also need to provide matchers to filter out applications for chain block in that request body, right?

Yes, we will need those `matchers`.

> not sure what do you mean by "we can have both chain ID and the first deployment ID at the same time"

I meant by that way, we can have chain ID, deployment objects, commands in the same API. Then the first deployment can have its needed chain ID seamlessly.

---

What I am concerned about is how to set the chain ID for the first deployment object. In the case of doing with separate RPCs, the flow will be like this, right?

- Piped generates chain ID locally
- Piped requests to create a new Deployment with the generated chain ID
- Piped requests to create a new Chain (commands, etc...)

IMO, generating the chain ID locally is not good. A Piped may send a chain ID of other projects (accidentally or intentionally).

So I think it would be better to avoid that. Another idea is reordering the above steps to

- Piped requests to create a new Chain (commands, etc..) and chain ID is returned
- Piped requests to create a new Deployment with the received chain ID

Reply:
Yeh, in that case, we have to use the chain id to find the chain and add deployment ref to chain: we have 2 different possible cases:
 + the common case: the creating deployment come by pipedx when it's triggered by chain_sync_app command, and besides with the chain id, we also need block index to add the deployment to the right block
 + the uncommon case: the creating deployment is the first deployment of the chain, in this case, we have to add a way to separate it with others, maybe by check for `matcher` existed in the request body or not, or specified block_index = 0 in the request body

> Piped requests to create a new Chain (commands, etc..) and chain ID is returned
One minor point for this is, at the time when we make commands chain_sync_app for applications in the chain, the first application deployment is not yet in the block[0] of the deployment chain model, if pipedx get the command and ask (via RPC) controlplane should it handle its command, we don't have any "previous" stage to determine. To avoid that, we have to ensure commands only be created after the first deployment is available as chain block[0] by making it as currently or adding `matcher` to the CreateDeploymentRequest as optional.

If you're worried about the ChainId generated in the client (piped side), we have another way to avoid that. That is generate it in the controlplane (API side) and update this `req.FirstDeployment` [here](https://github.com/pipe-cd/pipe/pull/2815/files#diff-dc8d28c3b055fcad0f022406b744ce756e3be8c630b6c40aef12e026c6a58dcfR1061) before actually create/add it to the datastore. How do you think about that?

Reply:
Addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/4a4ec2147c4103c411e5fe56a46d72334dd553fa